internal/rpc/starknet/types_test.go

```diff
@@ -0,0 +1,276 @@
+package starknet
+
+import (
+	"errors"
+	"io"
+	"reflect"
+	"testing"
+
+	"github.com/NethermindEth/juno/pkg/felt"
+	gocmp "github.com/google/go-cmp/cmp"
+
+	"gotest.tools/assert"
+)
+
+func TestBlockId_UnmarshalJSON(t *testing.T) {
+	tests := []struct {
+		name string
+		data []byte
+		want BlockId
+		err  error
+	}{
+		{
+			name: "empty data",
+			data: []byte{},
+			want: BlockId{},
+			err:  io.EOF,
+		},
+		{
+			name: "empty string",
+			data: []byte(`""`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+		{
+			name: "negative number",
+			data: []byte(`-1`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+		{
+			name: "float number",
+			data: []byte(`1.1`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+		{
+			name: "zero string",
+			data: []byte(`"0"`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+		{
+			name: "zero number",
+			data: []byte(`0`),
+			want: BlockId{
+				value:  int64(0),
+				idType: blockIdNumber,
+			},
+			err: nil,
+		},
+		{
+			name: "positive number",
+			data: []byte(`1`),
+			want: BlockId{
+				value:  int64(1),
+				idType: blockIdNumber,
+			},
+			err: nil,
+		},
+		{
+			name: "pending tag",
+			data: []byte(`"pending"`),
+			want: BlockId{
+				idType: blockIdTag,
+				value:  "pending",
+			},
+			err: nil,
+		},
+		{
+			name: "latest tag",
+			data: []byte(`"latest"`),
+			want: BlockId{
+				idType: blockIdTag,
+				value:  "latest",
+			},
+			err: nil,
+		},
+		{
+			name: "block hash",
+			data: []byte(`"0x07d328a71faf48c5c3857e99f20a77b18522480956d1cd5bff1ff2df3c8b427b"`),
+			want: BlockId{
+				idType: blockIdHash,
+				value:  new(felt.Felt).SetHex("0x07d328a71faf48c5c3857e99f20a77b18522480956d1cd5bff1ff2df3c8b427b"),
+			},
+			err: nil,
+		},
+		{
+			name: "invalid block hash",
+			data: []byte(`"0x7d328a71faf48c5c3857e99f20a77b18522480956d1cd5bff1ff2df3c8b427b"`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+		{
+			name: "unexpected JSON type",
+			data: []byte(`{}`),
+			want: BlockId{},
+			err:  ErrInvalidBlockId,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var got BlockId
+			if err := got.UnmarshalJSON(tt.data); !errors.Is(err, tt.err) {
+				t.Errorf("BlockId.UnmarshalJSON() = %v, want %v", err, tt.err)
+			}
+			assert.DeepEqual(t, got, tt.want, gocmp.Comparer(func(x, y BlockId) bool {
+				xv := reflect.ValueOf(x)
+				yv := reflect.ValueOf(y)
+				if xv.IsZero() || yv.IsZero() {
+					return xv.IsZero() && yv.IsZero()
+				}
+				if x.idType != y.idType {
+					return false
+				}
+				switch x.idType {
+				case blockIdHash:
+					return x.value.(*felt.Felt).Equal(y.value.(*felt.Felt))
+				case blockIdTag:
+					return x.value.(string) == y.value.(string)
+				case blockIdNumber:
+					return x.value.(int64) == y.value.(int64)
+				default:
+					return false
+				}
+			}))
+		})
+	}
+}
```

Comment:
Good test cases!

Can you please explain why `assert.DeepEqaul` is not sufficient?

Reply:
`assert.DeepEqual` does not know how to compare `BlockId`s. Regardless, we could add an `Equal` method to `BlockId`, which is what `assert.DeepEqual` looks for.
```go
func (id BlockId) Equal(other BlockId) bool {
	xv := reflect.ValueOf(id)
	yv := reflect.ValueOf(other)
	if xv.IsZero() || yv.IsZero() {
		return xv.IsZero() && yv.IsZero()
	}
	if id.idType != other.idType {
		return false
	}
	switch id.idType {
	case blockIdHash:
		return id.value.(*felt.Felt).Equal(other.value.(*felt.Felt))
	case blockIdTag:
		return id.value.(string) == other.value.(string)
	case blockIdNumber:
		return id.value.(int64) == other.value.(int64)
	default:
		return false
	}
}
```
With this change, `assert.DeepEqual` alone should be sufficient. This is nice because we don't need to copy the custom comparison function every time we call `DeepEqual` on structs with `BlockId`s.

Reply:
Because it broke on unexported fields
![image](https://user-images.githubusercontent.com/44682532/187713489-68278b57-0ad7-4bd2-8e6f-aa5018656540.png)

Reply:
Done -> https://github.com/NethermindEth/juno/pull/377/commits/e7a67c1953fc6118db7cfbbf14b6bbb3ccb9fcc9